integration/src/test/java/org/mule/test/integration/exceptions/ErrorHandlerWorksWithTxTestCase.java

```diff
@@ -0,0 +1,91 @@
+import static org.junit.Assert.assertThat;
+
+import org.junit.After;
+import org.junit.Test;
+import org.junit.runners.Parameterized;
+import org.mule.functional.api.component.EventCallback;
+import org.mule.runtime.api.event.Event;
+import org.mule.runtime.core.api.MuleContext;
+import org.mule.runtime.core.api.event.CoreEvent;
+import org.mule.test.AbstractIntegrationTestCase;
+import org.mule.test.runner.RunnerDelegateTo;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+@RunnerDelegateTo(Parameterized.class)
+public class ErrorHandlerWorksWithTxTestCase extends AbstractIntegrationTestCase {
+
+  private static final String ERROR_HANDLER_REF = "errorHandlerName";
+  private static List<Thread> threads = new ArrayList<>();
+  private String config;
+
+  @Parameterized.Parameters(name = "{0} - {2}")
+  public static Collection<Object[]> data() {
+    return Arrays.asList(new Object[][] {
+        {"Local Error Handler", "org/mule/test/integration/exceptions/error-handler-tx-same-thread.xml",
+            "errorHandlerWithTestNonBlocking"},
+        {"Local Error Handler", "org/mule/test/integration/exceptions/error-handler-tx-same-thread.xml",
+            "errorHandlerWithProcessingTypeChange"},
+        {"Local Error Handler", "org/mule/test/integration/exceptions/error-handler-tx-same-thread.xml",
+            "errorHandlerWithNonBlockingOp"},
+        {"Local Error Handler", "org/mule/test/integration/exceptions/error-handler-tx-same-thread.xml",
+            "errorHandlerWithNonBlockingOpAndProcessingTypeChange"},
+        {"Global Error Handler", "org/mule/test/integration/exceptions/error-handler-tx-same-thread-global-err.xml",
+            "errorHandlerWithTestNonBlocking"},
+        {"Global Error Handler", "org/mule/test/integration/exceptions/error-handler-tx-same-thread-global-err.xml",
+            "errorHandlerWithProcessingTypeChange"},
+        {"Global Error Handler", "org/mule/test/integration/exceptions/error-handler-tx-same-thread-global-err.xml",
+            "errorHandlerWithNonBlockingOp"},
+        {"Global Error Handler", "org/mule/test/integration/exceptions/error-handler-tx-same-thread-global-err.xml",
+            "errorHandlerWithNonBlockingOpAndProcessingTypeChange"},
+    });
+  }
+
+  @Override
+  protected String getConfigFile() {
+    return config;
+  }
+
+  public ErrorHandlerWorksWithTxTestCase(String type, String config, String errorHandlerName) {
+    System.setProperty(ERROR_HANDLER_REF, errorHandlerName);
+    this.config = config;
+  }
+
+  @After
+  public void tearDown() {
+    System.clearProperty(ERROR_HANDLER_REF);
+    threads.clear();
+  }
+
+  @Test
+  public void doesNotChangeThread() throws Exception {
+    Event event = flowRunner("flowWithTx").run();
+
+    assertThat(event.getMessage().getPayload().getValue(), is("zaraza"));
+    assertThat(threads, hasSize(2));
+    assertThat(threads.get(0), is(threads.get(1)));
```

Comment:
in a non-blocking world, the callback task may execute in the same thread. Set something in the thread local and assert against that instead of relying just in the thread identity.